File: 104. MaximumDepthofBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    if (!root) {
        return 0;
    }
    return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
};

var maxDepth = function(root) {
    let depth = 0;
    if (root) {
        const stack = [root];
        let count = stack.length;
        let cursor = 0;
        while(stack.length) {
            const node = stack.shift();
            cursor++;
            if (node.left) {
                stack.push(node.left);
            }
            if (node.right) {
                stack.push(node.right);
            }
            if (cursor === count) {
                depth++;
                count = stack.length;
                cursor = 0;
            }
        }
    }
    return depth;
}

File: Golang/Tree/104.Maximum-depth-of-binary-tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func maxDepthRecursive(root *TreeNode) int {
	var recursive func(node *TreeNode) int
	recursive = func(node *TreeNode) int {
		if node != nil {
			return 1 + max(recursive(node.Left), recursive(node.Right))
		}
		return 0
	}
	return recursive(root)
}

func maxDepthLayer(root *TreeNode) int {
	ret := 0
	if root != nil {
		stack := []*TreeNode{root}
		for len(stack) > 0 {
			children := []*TreeNode{}
			for i := 0; i < len(stack); i++ {
				node := stack[i]
				if node.Left != nil {
					children = append(children, node.Left)
				}
				if node.Right != nil {
					children = append(children, node.Right)
				}
			}
			ret += 1
			stack = children
		}
	}
	return ret
}
